Add option to set parallel test worker count to the physical core count of the machine

bogdanvlviv opened the pull request:
Also, use the physical core count of the machine as
the default number of workers, and generate the `test_helper.rb` file
with `parallelize(workers: :number_of_processors)`

Closes #34734

bogdanvlviv committed:
Add option to set parallel test worker count to the physical core count of the machine

Also, use the physical core count of the machine as
the default number of workers, and  generate the `test_helper.rb` file
with `parallelize(workers: :number_of_processors)`

Closes #34734

kaspth commented on activesupport/lib/active_support/test_case.rb:
```diff
@@ -12,6 +12,7 @@
 require "active_support/testing/time_helpers"
 require "active_support/testing/file_fixtures"
 require "active_support/testing/parallelization"
+require "concurrent"
```
Do we need to require the whole thing here?

bogdanvlviv replied:
Seems like a good idea just to require `"concurrent/utility/processor_counter"`

bogdanvlviv commented:
I think [tests failed](https://travis-ci.org/rails/rails/jobs/469281088#L2352) because the physical core count of the Travis' is 1.
I could reproduce those failures on my VM only by running those tests with `PARALLEL_WORKERS=1`.
I think we should change `parallelize(workers: :number_of_processors)` to `parallelize(workers: 2)` for those tests by [the similar way we do there](https://github.com/rails/rails/blob/master/railties/test/application/test_runner_test.rb#L545).

bogdanvlviv committed:
Ensure that we always test parallel testing

Since #34734 we decided to use the physical core count of the machine as
the default number of workers in the parallel testing, we need to
ensure that some tests use at least 2 workers because we could
run those tests on VM that has only 1 physical core.
It also fixes tests failures on the CI since Travis server we are using
has only one physical core.
See https://travis-ci.org/rails/rails/jobs/469281088#L2352

matthewd commented on railties/lib/rails/generators/rails/app/templates/test/test_helper.rb.tt:
```diff
@@ -5,9 +5,9 @@ require 'rails/test_help'
+  parallelize(workers: :number_of_processors, with: :threads)
 <%- else -%>
-  parallelize(workers: 2)
+  parallelize(workers: :number_of_processors)
```
If we're making the default an "it just works" value like this, does it make sense to still include it here? If most people don't need to override it, maybe this line can just be a plain `parallelize`?

bogdanvlviv replied:
I thought that it would be better to keep this in this form since it would be easier for users to figure out that their tests use the actual core count as a number of workes. But I don't have strong opinions. I think that we should either simplify this to `parallelize` and `parallelize(with: :threads)`(a few lines above), or extend it to `parallelize(workers: :number_of_processors, with: :processes)` to emphasize that test suit run use "fork approach".

dhh commented:
I like it spelling out as well 👍

> On Dec 19, 2018, at 00:55, Bogdan <notifications@github.com> wrote:
> 
> @bogdanvlviv commented on this pull request.
> 
> In railties/lib/rails/generators/rails/app/templates/test/test_helper.rb.tt:
> 
> >  <%- else -%>
> -  parallelize(workers: 2)
> +  parallelize(workers: :number_of_processors)
> I thought that it would be better to keep this in this form since it would be easier for users to figure out that their tests use the actual core count as a number of workes. But I don't have strong opinions. I think that we should either simplify this to parallelize and parallelize(with: :threads)(a few lines above), or extend it to parallelize(workers: :number_of_processors, with: :processes) to emphasize that test suit run use "fork approach".
> 
> —
> You are receiving this because you modified the open/close state.
> Reply to this email directly, view it on GitHub, or mute the thread.
